### PR TITLE
Fix leg generation to include alerts associated with the ending stop

### PR DESCRIPTION
**Pull request by evansiroky**
This change makes it so that alerts that are associated with the to stop of a leg are able to be properly included in the leg information.

### CODE REVIEW

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java [hunk 1]
```diff
@@ -871,7 +894,10 @@ private static void addPlaces(Leg leg, State[] states, Edge[] edges, boolean sho
 
         leg.from = makePlace(states[0], firstVertex, edges[0], firstStop, tripTimes, requestedLocale);
         leg.from.arrival = null;
-        leg.to = makePlace(states[states.length - 1], lastVertex, null, lastStop, tripTimes, requestedLocale);
+        int lastTransitStateIndex = states[states.length - 1].backEdge instanceof TransitBoardAlight
+            ? states.length - 2
+            : states.length - 1;
+        leg.to = makePlace(states[lastTransitStateIndex], lastVertex, null, lastStop, tripTimes, requestedLocale);
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java]: I would refactor the three instances of computing `lastTransitStateIndex` getting the last transit state into 
```java
State lastTransitState = getLastTransitState();
```
and replace all the instances of `states[lastTransitStateIndex]` with that variable.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Refactor the code for getting the last transit state.

**Review by landonreed — APPROVED** [review]
Agree with @binh-dam-ibigroup's suggestion to factor out the repeated code (to get either last transit state index or last state itself), but otherwise looks good.

**Review by binh-dam-ibigroup — APPROVED** [review]

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java [hunk 2]
```diff
@@ -901,6 +916,19 @@ private static void addPlaces(Leg leg, State[] states, Edge[] edges, boolean sho
         }
     }
 
+    /**
+     * In some leg states, the final state might be a TransitBoardAlight edge type, where instead we may want a pattern
+     * edge instead. This method will find and return the correct state to use for calculating items for a leg.
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java]: typo: duplicate "instead"

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java [hunk 3]
```diff
@@ -329,6 +330,18 @@ public static CoordinateArrayListSequence makeCoordinates(Edge[] edges) {
                 if (backMode != TraverseMode.LEG_SWITCH) {              // Start of leg switch
                     legIndexPairs[1] = i;
                 } else if (forwardMode != TraverseMode.LEG_SWITCH) {    // End of leg switch
+                    // if the start edge of the first state in the current leg index pair is a TransitBoardAlight edge
+                    // and the edge of the state just after the current leg index pair is a TransitBoardAlight edge,
+                    // then make that index be the final state of the current leg. This is to ensure that alerts
+                    // associated with TransitBoardAlight vertices are able to be found for legs.
+                    if (
+                        states[legIndexPairs[0]].getBackMode() != null &&
+                        states[legIndexPairs[0]].backEdge instanceof TransitBoardAlight &&
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java:339]: Is there not some common code in this conditional that can be factored out into a shared method used by `getLastOnboardTransitState`?

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java [hunk 4]
```diff
@@ -935,7 +963,7 @@ private static Place makePlace(State state, Vertex vertex, Edge edge, Stop stop,
             place.platformCode = stop.getPlatformCode();
             place.zoneId = stop.getZoneId();
             place.stopIndex = ((OnboardEdge) edge).getStopIndex();
-            if (endOfLeg) place.stopIndex++;
+            if (endOfLeg && !(edge instanceof TransitBoardAlight)) place.stopIndex++;
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java:968]: Apologies, missed this earlier - but this probably needs a comment?

**Review by landonreed — APPROVED** [review]
Changes look good, but there's some minor residual stuff that could be addressed.

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java (hunk 3 above)
Comment (evansiroky) [reply src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java:339]: No, not here. The `getLastOnboardTransitState` operates on data that is calculated after this code is executed and splits up the states.

**Pull request merged**